### PR TITLE
Customize notes for merge requests

### PR DESCRIPTION
**Pull request by sh0ked**
1. Sending message on result of a build as Post-build action.
2. You can set custom notes for success, failure and/or abort.
3. Jenkins environment variables interpolation is working in custom notes.

### CODE REVIEW

**Comment by coder-hugo** [comment]
I'd like to see this functionality in a post-build step. This will keep the UI and the code cleaner. If you want to adapt your code I suggest you to wait til I've finished my code cleanups of the plugin. This should happen hopefully within the next 2 weeks.

**Comment by omehegan** [comment]
@sh0ked if you can rebase this with the current master and refactor as @coder-hugo suggested, we'd like to get it merged!

This should also fix https://github.com/jenkinsci/gitlab-plugin/issues/168

**Comment by omehegan** [comment]
@sh0ked also, note that we need to avoid the use of AbstractBuild and AbstractProject in order to support Pipeline jobs. See https://github.com/jenkinsci/pipeline-plugin/blob/master/DEVGUIDE.md

**Comment by sh0ked** [comment]
@omehegan @coder-hugo Hi! Please check this merge request.

**Comment by skyline-gleb** [comment]
:+1:

**Comment by z00sts** [comment]
:+1:

File: src/main/java/com/dabsquared/gitlabjenkins/gitlab/api/GitLabApi.java [hunk 1]
```diff
@@ -1,9 +1,6 @@
 package com.dabsquared.gitlabjenkins.gitlab.api;
 
-import com.dabsquared.gitlabjenkins.gitlab.api.model.Branch;
-import com.dabsquared.gitlabjenkins.gitlab.api.model.BuildState;
-import com.dabsquared.gitlabjenkins.gitlab.api.model.Project;
-import com.dabsquared.gitlabjenkins.gitlab.api.model.User;
+import com.dabsquared.gitlabjenkins.gitlab.api.model.*;
```
Comment (coder-hugo) [inline src/main/java/com/dabsquared/gitlabjenkins/gitlab/api/GitLabApi.java:3]: Please do not use wildcard imports.

File: src/main/java/com/dabsquared/gitlabjenkins/publisher/GitLabMessagePublisher.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
+package com.dabsquared.gitlabjenkins.publisher;
+
+import com.dabsquared.gitlabjenkins.cause.GitLabWebHookCause;
+import com.dabsquared.gitlabjenkins.connection.GitLabConnectionProperty;
+import com.dabsquared.gitlabjenkins.gitlab.api.GitLabApi;
+import hudson.Extension;
+import hudson.Launcher;
+import hudson.Util;
+import hudson.model.*;
```
Comment (coder-hugo) [inline src/main/java/com/dabsquared/gitlabjenkins/publisher/GitLabMessagePublisher.java:9]: Please do not use wildcard imports.

File: src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java [hunk 3]
```diff
@@ -115,6 +116,9 @@ public static void migrateJobs() throws IOException {
                     if (trigger.addCiMessage) {
                         project.getPublishersList().add(new GitLabCommitStatusPublisher());
                     }
+                    if (trigger.addNoteOnMergeRequest) {
```
Comment (coder-hugo) [inline src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java:119]: This migration won't be triggered for anybody who already uses version 1.2.x of the plugin as for them the property jobsMigrated of the GitLabPushTrigger is already set to true. So this needs to be placed in a separate "migration block".

**Comment by coder-hugo** [comment]
The Notifier will only work for Freestyle jobs to support pipeline jobs as well we also need a WorkFlow step like the [GitLabCommitStatusStep](https://github.com/jenkinsci/gitlab-plugin/blob/master/src/main/java/com/dabsquared/gitlabjenkins/workflow/GitLabCommitStatusStep.java) for the merge request notes.

**Comment by ttsiebzehntt** [comment]
Hey, I would really like to see this merged (because of #168). Is only the WorkFlow step missing? Can I help somehow?

**Comment by kilpatty** [comment]
Anything I can do to help here? I would love to see this merged as well.

**Comment by sh0ked** [comment]
@ttsiebzehntt @Skilgarriff Guys, yes, help me. Now i have no time. This PR requires workflow step support. May be requires to change GitLabMessagePublisher class, move common code to separate class.

**Pull request closed without merging**